### DISCUSSION TITLE
Which is best language between Golang and NodeJS ?

### DISCUSSION THREAD

**Discussion by ambitious922** [community/community#58317] [Discussions] [answered]
### Select Topic Area

Question

### Body

NodeJS has been on the market for over 11 years. Since golang was released in the year 2009, it makes it just as old. Does that make a difference? For starters, NodeJS is written on C, C++, and JavaScript, while golang is fully based on C. It might seem that there’s not a huge difference, but they both contain a dozen features differentiating them from each other.
Because of golang’s support of concurrency, high performances, light-weight, speed, and memory bound, it’s what most developers might choose to work with. Yet, both golang and NodeJS yield the same result. NodeJS gains popularity day by day because of its JavaScript background, and it’s known that JavaScript is already massively popular among the dev community. But golang being static-typed, robust, and very secure, it might be hard to choose one of the two languages (or, between the language and the runtime environment), so let’s break it down.

**Comment by fourfor4** [comment, marked as answer]
**Here’s something we’ve scraped from the Golang documentation website:**
“Go is expressive, concise, clean, and efficient. Its concurrency mechanisms make it easy to write programs that get the most out of multicourse and networked machines, while its novel type system enables flexible and modular program construction. Go compiles quickly to machine code yet has the convenience of garbage collection and the power of run-time reflection. It’s a fast, statically typed, compiled language that feels like a dynamically typed, interpreted language.”

**While NodeJS’ documentation says otherwise:**
“As an asynchronous event-driven JavaScript runtime, Node.js is designed to build scalable network applications.Node.js is similar in design to, and influenced by, systems like Ruby’s [Event Machine](https://github.com/eventmachine/eventmachine) and Python’s [Twisted](https://twistedmatrix.com/trac/). Node.js takes the event model a bit further. It presents an [event loop](https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick/) as a runtime construct instead of as a library. In other systems, there is always a blocking call to start the event loop. Typically, behavior is defined through callbacks at the beginning of a script, and at the end, a server is started through a blocking call likeEventMachine::run(). In Node.js, there is no such start-the-event-loop call. Node.js simply enters the event loop after executing the input script. Node.js exits the event loop when there are no more callbacks to perform. This behavior is like browser JavaScript — the event loop is hidden from the user.”


**Both Nodejs and Golang brag about their concurrency machinery, but what’s concurrency? And which of the both have it better?**

Concurrency, in programming, is “algorithms”, that are executed simultaneously without affecting the end result.

If you’ve read both of these long documentation quoted paragraphs you’ll understand that NodeJS uses a single-threaded, event-callback mechanism that isn’t as “concurrent” as they brag about it, and each function is actually executed in a linear order. On the other hand, Golang has coroutines called “goroutines”, which handle concurrency. Goroutines are functions that execute independently or simultaneously with other goroutines to run your program with an actual concurrent mechanism. Thus, proven Golang is executing your functions concurrently. But again, due to thousands of goroutines used to yield your output concurrently, there is a lot, and I mean a lot of memory usage. But that is the only reason why golang is built for highly scalable and large applications.

[Need a good hosting solution? Try Hostgator!](https://partners.hostgator.com/bytestobits)

Error Handling. With its old-fashioned exception handling, NodeJS handles errors implicitly by wrapping around a try and catch. Golang handles errors explicitly. Although it may look redundant to people new to golang, this actually helps a lot, because it returns the errors and exhibits a First Come First Serve kind of appearance.

With NodeJS you can do both frontend and backend at the same time. This is a huge plus point, but again NodeJS supports asynchronous programming, which not everyone is familiar with, and golang is synchronous by default.

Developers, in particular, people with a lot of knowledge in both languages have actually picked golang instead of NodeJS for the above reasons, and you can read TJ Holowaychuk’s experience and what he migrated to, [here](https://medium.com/code-adventures/farewell-node-js-4ba9e7f3e52b).
